### PR TITLE
Add my endpoints

### PR DESCRIPTION
**Pull request by Hmiku8338**

### CODE REVIEW

File: e621/endpoints/endpoints.py [hunk 1]
```diff
@@ -1,10 +1,42 @@
...
+    def _default_get(self, identifier: Any, **kwargs: Any) -> Model:
+        return self._model.from_response(self._api.session.get(f"{self._url}/{identifier}", **kwargs), self._api)
+
+    def _default_search(
+        self, params: Dict[str, Any], limit: Optional[int], page: int = 1, ignore_pagination=False
+    ) -> List[Model]:
+        # ? Seems like a potential source of a bug
+        params.update({"limit": limit, "page": page})
+        if ignore_pagination:
+            return self._model.from_list(
```
Comment (eoan-ermine) [inline e621/endpoints/endpoints.py:57]: Subtraction from None if limit is None: [session.py:47](https://github.com/PatriotRossii/e621-py/blob/f540766d03a34671beda9b602cc638f6feb767be/e621/session.py#L47)
Was that intended?

File: e621/base_model.py [hunk 2]
```diff
@@ -1,7 +1,51 @@
...
+    @classmethod
+    def from_response(
+        cls,
+        response: requests.Response,
+        api: "E621API",
+        expect: Union[Type[Dict[Any, Any]], Type[List[Any]]] = dict,
+    ) -> Union[Self, List[Self]]:
+        json: Union[Dict[Any, Any], List[Any]] = response.json()
+        # {"post": {<post_info>}} or {"posts": [{<post_info>}, ...]}
+        if isinstance(json, dict) and len(cls.schema()["required"]) > len(json) and len(json) == 1:
```
Comment (eoan-ermine) [inline e621/base_model.py:46]: Why do we need `len(cls.schema()["required"]) > len(json)` here?

File: e621/endpoints/your_endpoints.py [hunk 3]
```diff
@@ -1,14 +1,181 @@
...
+    def me(self) -> AuthenticatedUser:
+        return AuthenticatedUser.from_response(self._api.session.get(f"users/{self._api.username}"), self._api)
 
+    def get(self, user_identifier: Union[UserID, UserName]) -> User:
+        return self._default_get(user_identifier)
 
-class PostFlags(BaseEndpoint):
-    pass
+    def search(self, limit: Optional[int] = None, page: int = 1, ignore_pagination: bool = False) -> List[User]:
+        return self._default_search({}, limit, page, ignore_pagination)
```
Comment (eoan-ermine) [inline e621/endpoints/your_endpoints.py]: Subtraction from None if limit is None, see [endpoints.py:29](https://github.com/PatriotRossii/e621-py/pull/3/files/f540766d03a34671beda9b602cc638f6feb767be#diff-4c6ac243b85e43ee943086f02de913d3074ff8ed960e86bcfbb877bf77d41af5)
Similarly, all other search() methods in endpoints
Was that intended?

File: e621/api.py [hunk 4]
```diff
@@ -50,27 +36,15 @@ def __init__(
...
-        return [p for p in posts if not p.all_tags & self.blacklist]
-
-    def get_post(self, post_id: int) -> EnrichedPost:
-        return EnrichedPost(**self.session.get(f"posts/{post_id}").json())
-
-    def get_pool(self, pool_id: int) -> EnrichedPool:
-        return EnrichedPool(**self.session.get(f"pools/{pool_id}").json(), _e6api=self)
+        self.users = endpoints.Users(self)
+
+    # ? This looks weird in an API class
```
Comment (eoan-ermine) [inline e621/api.py]: Yeah, it is. Can't it be moved to `AuthenticatedUser` or something?

File: e621/api.py [hunk 5]
```diff
@@ -50,27 +22,23 @@ def __init__(
...
+        self.tag_implications = endpoints.TagImplications(self)
+        self.bulk_update_requests = endpoints.BulkUpdateRequests(self)
+        self.blips = endpoints.Blips(self)
+        self.takedowns = endpoints.Takedowns(self)
+        self.user_feedbacks = endpoints.UserFeedbacks(self)
+        self.forum_topics = endpoints.ForumTopics(self)
+        self.post_sets = endpoints.PostSets(self)
+
+
+class E926(E621):
```
Comment (eoan-ermine) [inline e621/api.py:67]: Shouldn't classes E621 and E926 inherit from some abstract class?
E926 neither is a E621 nor extends it.

File: scripts/generate_models.py [hunk 6]
```diff
@@ -4,74 +4,85 @@
...
-                        "json",
-                        "--reuse-model",
-                        "--class-name",
-                        endpoint.replace("_", " ").title().replace(" ", ""),
-                    ]
-                )
-                api_endpoint_models.extend(normalize_models(temp_stdout.getvalue()))
+            json_template = e6api.session.get(endpoint, params={"limit": 500}).json()
+            all_jsons[endpoint] = json_template
+        final_json = (tempdir / "whatever").with_suffix(".json")
```
Comment (eoan-ermine) [inline scripts/generate_models.py]: "whatever" is really nice, but "models" or smth would be more appropriate

File: e621/endpoints/your_endpoints.py [hunk 7]
```diff
@@ -1,14 +1,343 @@
...
+            raise ValueError("Cannot access E621API.user for a non-authenticated user")
+        return AuthenticatedUser.from_response(self._api.session.get(f"users/{self._api.username}"), self._api)
+
+    def get(self, user_identifier: Union[UserID, UserName]) -> User:
+        return self._default_get(user_identifier)
+
+
+class PostVersions(EmptySearcher[PostVersion]):
+    _model = PostVersion
+    # search[updater_name]
```
Comment (eoan-ermine) [inline e621/endpoints/your_endpoints.py]: Implement all these shit

**Review by eoan-ermine — CHANGES_REQUESTED** [review]

File: e621/api.py (hunk 5 above)
Comment (Hmiku8338) [reply e621/api.py:67]: Why would we need an abstract class here? Not theoretically, but in terms of scaling/functionality.

And E926 **is** E621. It has all the same endpoints and the same functionality all around. E926 is just E621 with rating:safe being the only option.
image

File: e621/endpoints/endpoints.py (hunk 1 above)
Comment (Hmiku8338) [reply e621/endpoints/endpoints.py:57]: No. It is a bug. Thanks!

File: e621/endpoints/your_endpoints.py (hunk 7 above)
Comment (Hmiku8338) [reply e621/endpoints/your_endpoints.py]: Yass queen

File: e621/base_model.py (hunk 2 above)
Comment (Hmiku8338) [reply e621/base_model.py:46]: This line handles cases where return type is {"tags": []} for empty response and [...] for non-empty response. This specific expression you mentioned makes sure that we don't get it wrong when, for example, the model has only one field and that exact field is returned. We would handle that case incorrectly without that check.

**Pull request merged**